Allow adding referencing compute tags DataBox

nilsdeppe opened the pull request:
## Proposed changes

The use case for this is storing a pointer to the ConstGlobalCache, and then
having compute tags that are used to retrieve things stored in the
ConstGlobalCache using the base tag mechanism (OptionTags will be pure base
tags).

This doesn't add anything from the GlobalCache to the evolution DataBox, but adds the ability to do them to the DataBox. The `OptionTags` that we want to be able to retrieve from the DataBox will need to inherit from `db::BaseTag`.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Allow adding referencing compute tags DataBox

The use case for this is storing a pointer to the ConstGlobalCache, and then
having compute tags that are used to retrieve things stored in the
ConstGlobalCache using the base tag mechanism (OptionTags will be pure base
tags).

hrueter commented on src/DataStructures/DataBox/Deferred.hpp:
```diff
@@ -202,12 +265,31 @@ class deferred_assoc_state : public assoc_state<Rt> {
+ *
+ * \warning If passed a lazy function that returns a `const T& t` (a const
+ * lvalue reference) the underlying data stored is actually a pointer to `t`,
+ * which will be deferenced upon retrieval. This could lead to a dangling
```
do you mean "dereferenced"?

hrueter commented on src/Parallel/ConstGlobalCache.hpp:
```diff
@@ -210,6 +213,34 @@ auto get(const ConstGlobalCache<Metavariables>& cache) noexcept -> const
+/// recommended way for compute tags to retrieve data out of the global cache.
+template <class CacheTag>
+struct GlobalCache : CacheTag, db::ComputeTag {
+  static std::string name() noexcept { return "Uh Oh"; }
```
Why this tag name?

nilsdeppe replied:
So I'd remember to change it... Clearly I failed and you succeeded :) :+1:

nilsdeppe commented:
- Rebased on develop
- Addressed comments
- Added test for `std::unique_ptr` in the options since we return a `const &` of the underlying type, not the unique_ptr

wthrowe commented on src/DataStructures/DataBox/Deferred.hpp:
```diff
@@ -184,10 +184,73 @@ class deferred_assoc_state : public assoc_state<Rt> {
+    ERROR("Cannot mutate a compute tag.");
+    // clang-tidy: no const_cast (we need to do something to override the mutate
+    // function, which cannot be called for compute items anyway.)
+    return const_cast<Rt&>(*t_);  // NOLINT
```
You don't need to return after an ERROR, so you can just leave this out and avoid the NOLINT.

nilsdeppe replied:
Yep, done! Thanks :)

kidder reviewed (dismissed):
@wthrowe this has been updated

fmahebert commented on src/DataStructures/DataBox/Deferred.hpp:
```diff
@@ -184,10 +184,70 @@ class deferred_assoc_state : public assoc_state<Rt> {
+// return value as pointer to the data we actually want to have visible to us
+// when we retrieve the data. The `get` function dereferences the pointer we
+// store so that we have a reference to work with when retrieving the stored
+// data reference. Dereferencing the pointer ensures that all functions that use
```
I am finding the previous two sentences (from "We treat the return value...") difficult to understand. Is this rephrasing still accurate?

"We store a pointer to the returned data, i.e., a pointer to the data we actually want to retrieve from the DataBox. The `get` function dereferences the stored pointer, recovering a reference to the data to retrieve."

nilsdeppe replied:
Not quite because the data may not be in the DataBox (that's the whole point of this). The data can be stored anywhere, and the compute tag's function just returns a `const Rt&` to that data. However, since references are not rebindable, we have to store a pointer (which we can rebind).

fmahebert commented on src/DataStructures/DataBox/Deferred.hpp:
```diff
@@ -184,10 +184,70 @@ class deferred_assoc_state : public assoc_state<Rt> {
+// when we retrieve the data. The `get` function dereferences the pointer we
+// store so that we have a reference to work with when retrieving the stored
+// data reference. Dereferencing the pointer ensures that all functions that use
+// the DataBox will be able to take a `const T& t` regardless of where the data
```
"able to take a `const T& t` as input argument"

fmahebert commented on src/DataStructures/DataBox/Deferred.hpp:
```diff
@@ -202,12 +262,31 @@ class deferred_assoc_state : public assoc_state<Rt> {
  * Construction of a Deferred with an object followed by mutation:
  * \snippet Test_Deferred.cpp deferred_with_update
+ *
+ * \warning If passed a lazy function that returns a `const T& t` (a const
```
Should `T` be `Rt` so the documentation matches the implementation?

fmahebert commented on src/DataStructures/DataBox/Deferred.hpp:
```diff
@@ -202,12 +262,31 @@ class deferred_assoc_state : public assoc_state<Rt> {
+ *    compute tags and should return by value. If we find an actual use-case for
+ *    compute tags returning `const T&` where referencing behavior is undesired
+ *    then this design needs to be reconsidered. It is currently the least
+ *    breaking way to implement the desired functionality.
```
Actually state what "the desired functionality" is for clarity. Can you just refer back the previous item in this very list?

fmahebert commented on src/DataStructures/DataBox/Deferred.hpp:
```diff
@@ -184,10 +184,70 @@ class deferred_assoc_state : public assoc_state<Rt> {
   }
 };
+
+// Specialization to handle functions that return a `const T&`. We treat the
```
Should `T` be `Rt` so the documentation matches the implementation?

nilsdeppe replied:
It's not wrong, and if it avoids confusion all the better :) Making the change

nilsdeppe commented:
@wthrowe I updated the docs to address @fmahebert 's comments. Let us know if you're okay with it or not :)
